
# HG changeset patch
# User Hans de Goede <hdegoede@redhat.com>
# Date 1265106183 -3600
# Node ID 31edb32e4693d5cc0259ab3bea2ee8fa18aa85f2
# Parent  00c8a9985e2035ae6dce224d309fc0be25da6d8f
libv4l: skip false Pixart markers

From: Márton Németh <nm127@freemail.hu>

The byte sequence 0xff, 0xff, 0xff 0xff is not a real marker to skip, instead
it is one byte from the image and the following three 0xff bytes might belong
to a real marker. Modify pixart_fill_nbits() macro to pass the first 0xff byte
as an image data.

Priority: normal

Signed-off-by: Márton Németh <nm127@freemail.hu>
Signed-off-by: Hans de Goede <hdegoede@redhat.com>

--- a/v4l2-apps/libv4l/ChangeLog	Tue Feb 02 00:30:49 2010 +0100
+++ b/v4l2-apps/libv4l/ChangeLog	Tue Feb 02 11:23:03 2010 +0100
@@ -1,6 +1,8 @@
 libv4l-0.6.5
 ------------
 * Add more laptop models to the upside down devices table
+* Small fix to pixart jpeg decoding, this fixes the occasional corrupt
+  frame we used to get (Németh Márton)
 
 libv4l-0.6.4
 ------------
--- a/v4l2-apps/libv4l/libv4lconvert/tinyjpeg.c	Tue Feb 02 00:30:49 2010 +0100
+++ b/v4l2-apps/libv4l/libv4lconvert/tinyjpeg.c	Tue Feb 02 11:23:03 2010 +0100
@@ -339,10 +339,15 @@
 	    } \
 	    break; \
 	  case 0xff: \
-	    if (stream[1] == 0xff && (stream[2] < 7 || stream[2] == 0xff)) { \
-	      stream += 3; \
-	      c = *stream++; \
-	      break; \
+	    if (stream[1] == 0xff) { \
+		if (stream[2] < 7) { \
+		    stream += 3; \
+		    c = *stream++; \
+		    break; \
+		} else if (stream[2] == 0xff) { \
+		    /* four 0xff in a row: the first belongs to the image data */ \
+		    break; \
+		}\
 	    } \
 	    /* Error fall through */ \
 	  default: \

